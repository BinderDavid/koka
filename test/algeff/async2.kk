// --------------------------------------------------------
// Async demo: combine with ambiguity
// --------------------------------------------------------
effect amb {
  flip() : bool
}
 
val amb = handler {
  return x -> [x]
  flip()   -> resume(False) + resume(True)
}


function test() {
  println("what is your name?")
  val name = readline()
  if (flip()) then println("and your age?") else println("and your gender?")
  val rest = readline()
  println("hi " + name + ", you are " + rest)  
}

function main() {
  amb(test)
  () 
}