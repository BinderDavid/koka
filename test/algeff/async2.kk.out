.Op-flip : .op-flip
.Ops-flip : (flip : .op-flip) -> .ops-amb
.tag-amb : string
.tag-op-flip: string
amb : forall<a,e> (action : () -> <amb|e> a) -> e list<a>
flip : () -> amb bool
main : () -> <exn,std/async/async,console,ndet> ()
test : () -> <amb,exn,std/async/async,console,ndet> ()